src/libpgmoneta/io.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2024 The pgmoneta community
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <io.h>
+#include <logging.h>
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+ssize_t
+pgmoneta_write_file(int fd, void* buffer, size_t bytes)
+{
+   ssize_t bytes_written = 0;
+   int flags = fcntl(fd, F_GETFL);
+   int misalignment = (uintptr_t)buffer % BLOCK_SIZE;
+   int odd_bytes = bytes % BLOCK_SIZE;
+
+   if (flags & O_DIRECT)
+   {
+      if (misalignment != 0)
+      {
+         void* aligned_buf =
+            pgmoneta_aligned_malloc((bytes / BLOCK_SIZE + 1) * BLOCK_SIZE);
+         memcpy(aligned_buf, buffer, bytes);
+         buffer = aligned_buf;
+      }
+
+      if (odd_bytes != 0)
+      {
+         char* tmp_buf = (char*)buffer;
+         while (bytes_written + 8192 < bytes)
```

Comment:
Have you looked at https://eklitzke.org/efficient-file-copying-on-linux ?

Maybe for now we should just start with `#define WRITE_SIZE 8192`...

Reply:
Maybe try and see how different sizes affects your test setup

Reply:
Maybe an idea is to read the system buffer size into a `disk_buffer_size` and use that - default would be 8192

Reply:
128KB size buffer works best on my setup too. I have hardcoded this for now. I couldn't find a way to find the disk buffer size using libc calls. Maybe I can make a shell call to `blockdev` and find it.